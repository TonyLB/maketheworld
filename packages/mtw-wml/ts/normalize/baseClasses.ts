type TagType = 'Asset' |
    'Story' |
    'Character' |
    'Import' |
    'If' |
    'Description' |
    'Room' |
    'Feature' |
    'Image' |
    'Map' |
    'Exit' |
    'Variable' |
    'Computed' |
    'Action' |
    'Character'

export type NormalReference = {
    key: string;
    tag: TagType;
    index: number;
}

export type BaseAppearance = {
    contextStack: NormalReference[];
    contents: NormalReference[];
    location?: number[];
}

type NormalBase = {
    key: string;
}

export type NormalAsset = {
    tag: 'Asset';
    Story?: boolean;
    instance?: boolean;
    appearances: BaseAppearance[];
    fileName?: string;
    zone?: string;
} & NormalBase

export type NormalCharacterPronouns = {
    subject: string;
    object: string;
    reflexive: string;
    possessive: string;
    adjective: string;
}

export type NormalCharacter = {
    tag: 'Character';
    Name: string;
    Pronouns: NormalCharacterPronouns;
    FirstImpression: string;
    OneCoolThing: string;
    Outfit: string;
    fileName?: string;
    fileURL?: string;
    appearances: BaseAppearance[];
} & NormalBase

export type ComponentRenderItem = {
    tag: 'Link';
    targetTag: 'Feature' | 'Action';
    to: string;
    text: string;
} | {
    tag: 'String';
    value: string;
} | {
    tag: 'LineBreak';
} | {
    tag: 'Space';
} | {
    tag: 'Condition';
    if: string;
    dependencies: string[];
    contents: ComponentRenderItem[];
}

export type NormalDescriptionPayload = {
    type: 'Description';
    render?: ComponentRenderItem[];
}

export type NormalDescription = NormalDescriptionPayload & NormalBase

export type ComponentAppearance = Omit<NormalDescriptionPayload, 'type'> & BaseAppearance & {
    name?: ComponentRenderItem[];
    x?: number;
    y?: number;
}

type ComponentTypes = 'Room' | 'Feature'

export type NormalComponent = {
    tag: ComponentTypes;
    appearances: ComponentAppearance[];
    global?: boolean;
} & NormalBase

export type NormalRoom = NormalComponent & {
    tag: 'Room';
}

export type NormalFeature = NormalComponent & {
    tag: 'Feature';
}

type NormalImportMapping = {
    key: string;
    type: string;
}

export type NormalImport = {
    tag: 'Import';
    from: string;
    mapping: Record<string, NormalImportMapping>;
    appearances: BaseAppearance[];
} & NormalBase

export type NormalCondition = {
    tag: 'If';
    if: string;
    dependencies: string[];
    appearances: BaseAppearance[];
} & NormalBase

export type ImageAppearance = {
    display?: 'replace'
} & BaseAppearance

export type NormalImage = {
    tag: 'Image';
    fileURL?: string;
    appearances: ImageAppearance[];
} & NormalBase

export type MapAppearanceRoom = {
    conditions: {
        if: string;
        dependencies: string[]
    }[];
    key: string;
    x: number;
    y: number;
    location: number[];
}

type MapAppearanceImage = string

export type MapAppearance = {
    name: ComponentRenderItem[];
    images: MapAppearanceImage[];
    rooms: MapAppearanceRoom[];
} & BaseAppearance

export type NormalMap = {
    tag: 'Map';
    appearances: MapAppearance[];
} & NormalBase

type ExitAppearance = {
    name?: string;
} & BaseAppearance

export type NormalExit = {
    tag: 'Exit';
    to: string;
    from: string;
    name?: string;
    appearances?: ExitAppearance[];
} & NormalBase

export type NormalVariable = {
    tag: 'Variable';
    default: string;
    appearances: BaseAppearance[];
} & NormalBase

export type NormalComputed = {
    tag: 'Computed';
    src: string;
    dependencies: string[];
    appearances: BaseAppearance[];
} & NormalBase

export type NormalAction = {
    tag: 'Action';
    src: string;
    appearances: BaseAppearance[];
} & NormalBase

export type NormalItem = NormalAsset |
    NormalComponent |
    NormalImport |
    NormalImage |
    NormalMap |
    NormalExit |
    NormalVariable |
    NormalComputed |
    NormalAction |
    NormalCondition |
    NormalCharacter

export type NormalForm = Record<string, NormalItem>

export class WMLNormalizeError extends Error {
    constructor(message: string) {
        super(`WMLNormalize: ${message}`)
        this.name = 'WMLNormalizeError'
    }
}

export class NormalizeTagMismatchError extends WMLNormalizeError {
    constructor(message: string) {
        super(message)
        this.name = 'TagMismatchError'
    }
}

export class NormalizeKeyMismatchError extends WMLNormalizeError {
    constructor(message: string) {
        super(message)
        this.name = 'KeyTagMismatchError'
    }
}

export const isNormalExit = (arg: NormalItem): arg is NormalExit => (arg?.tag === 'Exit')
export const isNormalImage = (arg: NormalItem): arg is NormalImage => (arg?.tag === 'Image')
export const isNormalComponent = (arg: NormalItem): arg is NormalComponent => (['Room', 'Feature'].includes(arg?.tag))
export const isNormalRoom = (arg: NormalItem): arg is NormalRoom => (arg?.tag === 'Room')
export const isNormalFeature = (arg: NormalItem): arg is NormalFeature => (arg?.tag === 'Feature')
export const isNormalMap = (arg: NormalItem): arg is NormalMap => (arg?.tag === 'Map')
export function isNormalAsset(arg: NormalItem): arg is NormalAsset {
    return arg?.tag === 'Asset'
}
export const isNormalCharacter = (arg: NormalItem): arg is NormalCharacter => (arg?.tag === 'Character')
export const isNormalImport = (arg: NormalItem): arg is NormalImport => (arg?.tag === 'Import')
export const isNormalCondition = (arg: NormalItem): arg is NormalCondition => (arg.tag === 'If')
export const isNormalVariable = (arg: NormalItem): arg is NormalVariable => (arg.tag === 'Variable')
export const isNormalComputed = (arg: NormalItem): arg is NormalComputed => (arg.tag === 'Computed')
export const isNormalAction = (arg: NormalItem): arg is NormalAction => (arg.tag === 'Action')
