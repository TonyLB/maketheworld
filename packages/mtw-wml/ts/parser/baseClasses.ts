import {
    Token,
    TokenExpressionValue,
    TokenKeyValue,
    TokenLiteralValue,
    ParseStackTokenEntry
} from './tokenizer/baseClasses'

type ParseTagBase = {
    startTagToken: number;
    endTagToken: number;
}

export type ParseLiteralLegalContents = ParseStringTag
type ParseValueBase = {
    value: string;
    contents: ParseLiteralLegalContents[];
} & ParseTagBase

export type ParseAssetLegalContents = ParseActionTag |
    ParseCommentTag |
    ParseComputedTag |
    ParseConditionTag |
    ParseExitTag |
    ParseFeatureTag |
    ParseImageTag |
    ParseImportTag |
    ParseMapTag |
    ParseRoomTag |
    ParseVariableTag |
    ParseWhitespaceTag

type ParseAssetBase = {
    key: string;
    fileName?: string;
    zone?: string;
    subFolder?: string;
    player?: string;
    contents: ParseAssetLegalContents[];
} & ParseTagBase

export type ParseAssetTag = {
    tag: 'Asset';
} & ParseAssetBase

export type ParseStoryTag = {
    tag: 'Story';
    instance: boolean;
} & ParseAssetBase

export type ParseNameTag = ParseTaggedMessageTag<"Name">

export type ParsePronounsTag = {
    tag: 'Pronouns';
    subject: string;
    object: string;
    possessive: string;
    reflexive: string;
    adjective: string;
} & ParseTagBase

export type ParseOneCoolThingTag = {
    tag: 'OneCoolThing';
} & ParseValueBase

export type ParseFirstImpressionTag = {
    tag: 'FirstImpression';
} & ParseValueBase

export type ParseOutfitTag = {
    tag: 'Outfit';
} & ParseValueBase

export type ParseImageTag = {
    tag: 'Image';
    key: string;
    fileURL?: string;
} & ParseTagBase

export type ParseCharacterLegalContents = ParseNameTag | ParsePronounsTag | ParseFirstImpressionTag | ParseOneCoolThingTag | ParseOutfitTag | ParseImageTag
export type ParseCharacterTag = {
    tag: 'Character';
    key: string;
    player?: string;
    fileName?: string;
    zone?: string;
    subFolder?: string;
    contents: ParseCharacterLegalContents[];
} & ParseTagBase

export type ParseVariableTag = {
    tag: 'Variable';
    key: string;
    default?: string;
} & ParseTagBase

export type ParseDependencyTag = {
    tag: 'Depend';
    on: string;
} & ParseTagBase

export type ParseComputedTag = {
    tag: 'Computed';
    key: string;
    src: string;
    dependencies: ParseDependencyTag[];
} & ParseTagBase

export type ParseActionTag = {
    tag: 'Action';
    key: string;
    src: string;
} & ParseTagBase

export type ParseUseTag = {
    tag: 'Use';
    key: string;
    as?: string;
    type?: string;
} & ParseTagBase

export type ParseImportLegalContents = ParseUseTag
export type ParseImportTag = {
    tag: 'Import';
    from: string;
    contents: ParseImportLegalContents[];
} & ParseTagBase

export type ParseConditionLegalContextTag = 'Asset' | 'Description' | 'Room' | 'Feature' | 'Map'

export type ParseConditionTagAssetContext = {
    tag: 'If';
    contextTag: 'Asset';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseAssetLegalContents[];
} & ParseTagBase

export type ParseConditionTagDescriptionContext = {
    tag: 'If';
    contextTag: 'Description';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseTaggedMessageLegalContents[];
} & ParseTagBase

export type ParseConditionTagRoomContext = {
    tag: 'If';
    contextTag: 'Room';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseRoomLegalContents[];
} & ParseTagBase

export type ParseConditionTagFeatureContext = {
    tag: 'If';
    contextTag: 'Feature';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseFeatureLegalContents[];
} & ParseTagBase

export type ParseConditionTagMapContext = {
    tag: 'If';
    contextTag: 'Map';
    if: string;
    dependencies: ParseDependencyTag[];
    contents: ParseMapLegalContents[];
} & ParseTagBase

export type ParseConditionTypeFromContextTag<T extends ParseConditionLegalContextTag> =
    T extends 'Description'
        ? ParseConditionTagDescriptionContext
        : T extends 'Room'
            ? ParseConditionTagRoomContext
            : T extends 'Feature'
                ? ParseConditionTagFeatureContext
                : T extends 'Map'
                    ? ParseConditionTagMapContext
                    : ParseConditionTagAssetContext

export type ParseConditionTag = ParseConditionTagAssetContext | ParseConditionTagDescriptionContext | ParseConditionTagRoomContext | ParseConditionTagFeatureContext | ParseConditionTagMapContext

export const isLegalParseConditionContextTag = (value: string): value is ParseConditionLegalContextTag => (['Asset', 'Description', 'Room', 'Feature', 'Map'].includes(value))
export const isParseConditionTagAssetContext = (value: ParseConditionTag): value is ParseConditionTagAssetContext => (value.contextTag === 'Asset')
export const isParseConditionTagDescriptionContext = (value: ParseConditionTag): value is ParseConditionTagDescriptionContext => (value.contextTag === 'Description')
export const isParseConditionTagRoomContext = (value: ParseConditionTag): value is ParseConditionTagRoomContext => (value.contextTag === 'Room')
export const isParseConditionTagFeatureContext = (value: ParseConditionTag): value is ParseConditionTagFeatureContext => (value.contextTag === 'Feature')
export const isParseConditionTagMapContext = (value: ParseConditionTag): value is ParseConditionTagMapContext => (value.contextTag === 'Map')

export const parseDifferentiatingTags: Record<ParseConditionLegalContextTag,  ParseTag["tag"][]> = {
    Asset: ['Exit', 'Feature', 'Room', 'If', 'Image', 'Map'],
    Description: ['If', 'Space', 'String', 'Link', 'br', 'Whitespace'],
    Room: ['If', 'Description', 'Name', 'Exit'],
    Feature: ['If', 'Description', 'Name'],
    Map: ['If', 'Image', 'Room', 'Name', 'Exit']
}

export type ParseExitTag = {
    tag: 'Exit';
    key?: string;
    name: string;
    to?: string;
    from?: string;
    contents: ParseStringTag[];
} & ParseTagBase

export type ParseLinkLegalContents = ParseStringTag | ParseWhitespaceTag
export type ParseLinkTag = {
    tag: 'Link';
    to: string;
    contents: ParseLinkLegalContents[];
} & ParseTagBase

export type ParseTaggedMessageLegalContents = ParseWhitespaceTag | ParseStringTag | ParseLinkTag | ParseLineBreakTag | ParseSpacerTag | ParseConditionTagDescriptionContext

export type ParseTaggedMessageTag<T extends string> = {
    tag: T;
    display?: 'before' | 'after' | 'replace';
    contents: ParseTaggedMessageLegalContents[];
} & ParseTagBase

export type ParseDescriptionTag = ParseTaggedMessageTag<"Description">

export type ParseLineBreakTag = {
    tag: 'br';
} & ParseTagBase

export type ParseSpacerTag = {
    tag: 'Space';
} & ParseTagBase

export type ParseRoomLegalContents = ParseDescriptionTag | ParseNameTag | ParseExitTag | ParseFeatureTag | ParseConditionTagRoomContext
export type ParseRoomTag = {
    tag: 'Room';
    key: string;
    global: boolean;
    display?: string;
    x?: number;
    y?: number;
    contents: ParseRoomLegalContents[];
} & ParseTagBase

export type ParseFeatureLegalContents = ParseDescriptionTag | ParseNameTag | ParseConditionTagFeatureContext
export type ParseFeatureTag = {
    tag: 'Feature';
    key: string;
    global: boolean;
    contents: ParseFeatureLegalContents[];
} & ParseTagBase

export type ParseMapLegalContents = ParseNameTag | ParseRoomTag | ParseImageTag | ParseExitTag | ParseConditionTagMapContext
export type ParseMapTag = {
    tag: 'Map';
    key: string;
    contents: ParseMapLegalContents[];
} & ParseTagBase

export type ParseStringTag = {
    tag: 'String';
    value: string;
} & ParseTagBase

export type ParseWhitespaceTag = {
    tag: 'Whitespace';
} & ParseTagBase

export type ParseCommentTag = {
    tag: 'Comment';
} & ParseTagBase

export type ParseTag = ParseAssetTag |
    ParseStoryTag |
    ParseNameTag |
    ParseCharacterTag |
    ParsePronounsTag |
    ParseOneCoolThingTag |
    ParseFirstImpressionTag |
    ParseOutfitTag |
    ParseImageTag |
    ParseVariableTag |
    ParseComputedTag |
    ParseDependencyTag |
    ParseActionTag |
    ParseUseTag |
    ParseImportTag |
    ParseConditionTag |
    ParseExitTag |
    ParseDescriptionTag |
    ParseLineBreakTag |
    ParseLinkTag |
    ParseRoomTag |
    ParseFeatureTag |
    ParseMapTag |
    ParseStringTag |
    ParseWhitespaceTag |
    ParseSpacerTag |
    ParseCommentTag

export type ParseLegalTag = ParseTag["tag"]

export const isParseLegalTag = (tag: string): tag is ParseLegalTag => ([
    'Asset',
    'Story',
    'Name',
    'Character',
    'Pronouns',
    'OneCoolThing',
    'FirstImpression',
    'Outfit',
    'Image',
    'Variable',
    'Computed',
    'Depend',
    'Action',
    'Use',
    'Import',
    'If',
    'Exit',
    'Description',
    'br',
    'Link',
    'Room',
    'Feature',
    'Map',
    'String',
    'Whitespace',
    'Space',
    'Comment'
].includes(tag))

export type ParseStackTagOpenPendingEntry = {
    type: 'TagOpenPending';
    tag: string;
    startTagToken: number;
}

export type ParseStackTagOpenEntry = {
    type: 'TagOpen';
    tag: ParseLegalTag;
    startTagToken: number;
    properties: Record<string, (TokenExpressionValue | TokenKeyValue | TokenLiteralValue | boolean)>;
}

export type ParseStackTagEntry<T extends ParseTag> = {
    type: 'Tag';
    tag: T;
}

export class ParseException extends Error {
    startToken: number
    endToken: number
    constructor(message: string, startToken: number, endToken: number) {
        super(message)
        this.name = 'ParseException'
        this.startToken = startToken
        this.endToken = endToken
    }
}

export const isParseTagDependency = (value: ParseTag): value is ParseDependencyTag => (value.tag === 'Depend')
export const isParseTagNesting = (value: ParseTag): value is (ParseRoomTag | ParseFeatureTag | ParseAssetTag | ParseStoryTag | ParseCharacterTag | ParseImportTag | ParseDescriptionTag | ParseConditionTag | ParseLinkTag | ParseMapTag | ParseExitTag | ParseNameTag | ParseFirstImpressionTag | ParseOneCoolThingTag | ParseOutfitTag) => (
    ['Room', 'Feature', 'Asset', 'Story', 'Character', 'Import', 'Description', 'If', 'Link', 'Map', 'Exit', 'Name', 'FirstImpression', 'OneCoolThing', 'Outfit'].includes(value.tag)
)
export const isParseExit = (value: ParseTag): value is ParseExitTag => (value.tag === 'Exit')
export const isParseRoom = (value: ParseTag): value is ParseRoomTag => (value.tag === 'Room')
export const isParseString = (value: ParseTag): value is ParseStringTag => (value.tag === 'String')

export type ParseStackEntry = ParseStackTagOpenPendingEntry | ParseStackTagOpenEntry | ParseStackTagEntry<ParseTag> | ParseStackTokenEntry<Token>

export const isParseStackTagOpenEntry = (value: ParseStackEntry): value is ParseStackTagOpenEntry => (value.type === 'TagOpen')

export type ParseTagFactory<T extends ParseTag> = (value: { open: ParseStackTagOpenEntry, context: ParseStackTagOpenEntry[]; contents: ParseTag[], endTagToken: number }) => ParseStackTagEntry<T>

export const parseTagDefaultProps: Record<string, string> = {
    If: 'if'
}
