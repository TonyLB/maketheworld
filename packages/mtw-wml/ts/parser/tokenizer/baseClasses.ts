import SourceStream from "./sourceStream";

export type TokenBase = {
    source: string;
    startIdx: number;
    endIdx: number;
}

export type TokenComment = {
    type: 'Comment'
} & TokenBase

export type TokenWhitespace = {
    type: 'Whitespace'
} & TokenBase

export type TokenExpressionValue = {
    type: 'ExpressionValue'
} & TokenBase

export type TokenKeyValue = {
    type: 'KeyValue';
    key: string;
} & TokenBase

export type TokenLiteralValue = {
    type: 'LiteralValue';
    value: string;
} & TokenBase

export type TokenProperty = {
    type: 'Property'
} & TokenBase

export type TokenBeginTagOpen = {
    type: 'BeginTagOpen'
} & TokenBase

export type TokenBeginTagClose = {
    type: 'BeginTagClose';
} & TokenBase

export type TokenEndTag = {
    type: 'EndTag';
} & TokenBase

export type TokenDescription = {
    type: 'Description';
} & TokenBase

export type TokenError = {
    type: 'Error';
    message: string;
} & TokenBase

export type Token = TokenComment |
    TokenWhitespace |
    TokenExpressionValue |
    TokenKeyValue |
    TokenLiteralValue |
    TokenProperty |
    TokenBeginTagOpen |
    TokenBeginTagClose |
    TokenEndTag |
    TokenDescription |
    TokenError

export type Tokenizer<T extends TokenBase> = {
    (sourceStream: SourceStream): T | TokenError | undefined
}

export const isTokenError = (item: TokenBase & { type: string }): item is TokenError => (item.type === 'Error')
