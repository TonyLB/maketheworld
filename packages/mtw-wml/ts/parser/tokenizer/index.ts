import {
    isTokenError,
    Token,
    TokenDescription,
    TokenTagClose,
    TokenTagOpenBegin,
    TokenTagOpenEnd,
    TokenWhitespace,
    TokenProperty,
    TokenBooleanProperty,
    TokenLiteralValue,
    TokenKeyValue,
    TokenExpressionValue
} from "./baseClasses"
import booleanPropertyTokenizer from "./boolean"
import expressionValueTokenizer from "./expression"
import keyValueTokenizer from "./key"
import literalValueTokenizer from "./literal"
import SourceStream from "./sourceStream"
import tagCloseTokenizer from "./tagClose"

import { tagOpenBeginTokenizer, tagOpenEndTokenizer } from "./tagOpen"
import tagPropertyTokenizer from "./tagProperty"
import whiteSpaceTokenizer from "./whiteSpace"
import { checkSubTokenizers } from "./utils"

enum TokenizerMode {
    Contents,
    Tag
}

export const tokenizer = (sourceStream: SourceStream): Token[] => {
    let mode: TokenizerMode = TokenizerMode.Contents
    let currentDescription: TokenDescription | undefined
    let returnValue: Token[] = []

    while(!sourceStream.isEndOfSource) {
        if (mode === TokenizerMode.Contents) {
            const checkSubTokens = checkSubTokenizers<TokenWhitespace | TokenTagClose | TokenTagOpenBegin>({
                sourceStream,
                subTokenizers: [whiteSpaceTokenizer, tagCloseTokenizer, tagOpenBeginTokenizer],
                callback: (token) => {
                    returnValue.push(token)
                    if (token.type === 'TagOpenBegin') {
                        mode = TokenizerMode.Tag
                    }
                }
            })
            if (checkSubTokens) {
                if (checkSubTokens.success === false) {
                    return [checkSubTokens.error]
                }
                continue
            }
            return [{
                type: 'Error',
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                source: sourceStream.lookAhead(1),
                message: 'Unexpected token'
            }]
        }
        else {
            const checkSubTokens = checkSubTokenizers<TokenTagOpenEnd | TokenWhitespace | TokenProperty | TokenBooleanProperty>({
                sourceStream,
                subTokenizers: [tagOpenEndTokenizer, whiteSpaceTokenizer, tagPropertyTokenizer, booleanPropertyTokenizer],
                callback: (token) => {
                    returnValue.push(token)
                    if (token.type === 'TagOpenEnd') {
                        mode = TokenizerMode.Contents
                    }
                    if (token.type === 'Property') {
                        //
                        // Consume any whitespace first...
                        //
                        const checkWhitespace = whiteSpaceTokenizer(sourceStream)
                        if (checkWhitespace) {
                            if (isTokenError(checkWhitespace)) {
                                return {
                                    success: false,
                                    error: checkWhitespace
                                }
                            }
                            returnValue.push(checkWhitespace)
                        }
                        //
                        // ... and then confirm that a value follows the key
                        //
                        const checkValue = checkSubTokenizers<TokenLiteralValue | TokenKeyValue | TokenExpressionValue>({
                            sourceStream,
                            subTokenizers: [literalValueTokenizer, keyValueTokenizer, expressionValueTokenizer],
                            callback: (token) => {
                                returnValue.push(token)
                            }
                        })
                        if (checkValue) {
                            if (checkValue.success === false) {
                                return checkValue
                            }
                        }
                        else {
                            return {
                                success: false,
                                error: {
                                    type: 'Error',
                                    source: sourceStream.lookAhead(1),
                                    startIdx: sourceStream.position,
                                    endIdx: sourceStream.position,
                                    message: 'Unexpected token'
                                }
                            }
                        }
                    }
                }
            })
            if (checkSubTokens) {
                if (checkSubTokens.success === false) {
                    return [checkSubTokens.error]
                }
                continue
            }
            return [{
                type: 'Error',
                source: sourceStream.lookAhead(1),
                startIdx: sourceStream.position,
                endIdx: sourceStream.position,
                message: 'Unexpected token'
            }]
        }
    }
    return returnValue
}

export default tokenizer
