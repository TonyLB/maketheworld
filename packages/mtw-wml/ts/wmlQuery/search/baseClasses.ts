import SourceStream from "../../parser/tokenizer/sourceStream"

export type SearchTokenWhitespace = {
    type: 'Whitespace';
}

type SearchTokenTagLegal = "Asset" |
    "Story" |
    "Character" |
    "Name" |
    "Pronouns" |
    "OneCoolThing" |
    "Outfit" |
    "FirstImpression" |
    "Condition" |
    "Layer" |
    "Exit" |
    "Description" |
    "Room" |
    "Feature" |
    "Map"

export type SearchTokenTag = {
    type: 'Tag';
    tag:  SearchTokenTagLegal;
}

export type SearchTokenString = {
    type: 'String';
    value: string;
}

type SearchTokenPropertyLegal = 'key' | 'to' | 'from'
export type SearchTokenProperty = {
    type: 'Property';
    value: SearchTokenPropertyLegal
}

export type SearchTokenFirst = {
    type: 'First';
}

export type SearchTokenNthChild = {
    type: 'NthChild';
    n: number;
}

export type SearchTokenGroupOpen = {
    type: 'GroupOpen';
}

export type SearchTokenGroupClose = {
    type: 'GroupClose';
}

export type SearchTokenComma = {
    type: 'Comma';
}

export type SearchToken = SearchTokenTag |
    SearchTokenString |
    SearchTokenProperty |
    SearchTokenFirst |
    SearchTokenNthChild |
    SearchTokenGroupOpen |
    SearchTokenGroupClose |
    SearchTokenComma |
    SearchTokenWhitespace

export type SearchTokenizer<T extends { type: string }> = {
    (sourceStream: SourceStream): T | undefined
}    

export type SearchParseTag = {
    type: 'Tag';
    tag: SearchTokenTagLegal;
}

export type SearchParseProperty = {
    type: 'Property';
    key: SearchTokenPropertyLegal;
    value: string;
}

export type SearchParseFirst = {
    type: 'First';
}

export type SearchParseNthChild = {
    type: 'NthChild';
    n: number;
}

export type SearchParseGroup = {
    type: 'Group';
    items: SearchParse[];
}

export type SearchParse = SearchParseTag |
    SearchParseProperty |
    SearchParseFirst |
    SearchParseNthChild |
    SearchParseGroup
