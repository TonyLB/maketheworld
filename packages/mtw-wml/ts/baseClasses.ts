import { ParseTag } from "./parser/baseClasses";

type SchemaNestingBase<T extends SchemaTag> = {
    contents: T[];
}

export type LegalAssetContents = SchemaActionTag | SchemaComputedTag | SchemaConditionTag | SchemaExitTag | SchemaFeatureTag | SchemaImageTag | SchemaImportTag | SchemaMapTag | SchemaRoomTag | SchemaVariableTag
export type LegalConditionContents = SchemaActionTag | SchemaComputedTag | SchemaConditionTag | SchemaExitTag | SchemaFeatureTag | SchemaImageTag | SchemaImportTag | SchemaMapTag | SchemaRoomTag | SchemaVariableTag

type SchemaAssetBase = {
    key: string;
    fileName?: string;
    zone?: string;
    subFolder?: string;
    player?: string;
} & SchemaNestingBase<LegalAssetContents>

export type SchemaAssetTag = {
    tag: 'Asset';
} & SchemaAssetBase

export type SchemaStoryTag = {
    tag: 'Story';
    instance: boolean;
} & SchemaAssetBase

export type SchemaImageTag = {
    tag: 'Image';
    key: string;
    fileURL?: string;
}

type SchemaPronouns = {
    subject: string;
    object: string;
    possessive: string;
    adjective: string;
    reflexive: string;
}

export type SchemaCharacterTag = {
    tag: 'Character';
    player?: string;
    name: string;
    pronouns: SchemaPronouns;
    firstImpression?: string;
    oneCoolThing?: string;
    outfit?: string;
    image?: SchemaImageTag;
}

export type SchemaVariableTag = {
    tag: 'Variable';
    key: string;
    default?: string;
}

export type SchemaComputedTag = {
    tag: 'Computed';
    key: string;
    src: string;
    dependencies: string[];
}

export type SchemaActionTag = {
    tag: 'Action';
    src: string;
}

export type SchemaUseTag = {
    tag: 'Use';
    key: string;
    as?: string;
    type?: string;
}

type SchemaImportMapping = {
    key: string;
    type: 'Room' | 'Feature' | 'Variable' | 'Computed' | 'Action' | 'Map'
}

export type SchemaImportTag = {
    tag: 'Import';
    from: string;
    mapping: Record<string, SchemaImportMapping>;
}

//
// TODO:  Figure out how to define limits on schemaConditionTag contents
// without causing circular reference
//
export type SchemaConditionTag = {
    tag: 'Condition';
    if: string;
    dependencies: string[];
} & SchemaNestingBase<any>

export type SchemaExitTag = {
    tag: 'Exit';
    key: string;
    name: string;
    to: string;
    from: string;
}

export type SchemaLinkTag = {
    tag: 'Link';
    to: string;
} & SchemaNestingBase<SchemaStringTag>

export type SchemaDescriptionTag = {
    tag: 'Description';
    display: 'before' | 'after' | 'replace';
    spaceBefore: boolean;
    spaceAfter: boolean;
} & SchemaNestingBase<SchemaStringTag | SchemaLinkTag | SchemaLineBreakTag>

export type SchemaLineBreakTag = {
    tag: 'br';
}

export type SchemaNameTag = {
    tag: 'Name';
    name: string;
}

export type SchemaRoomTag = {
    tag: 'Room';
    key: string;
    global: boolean;
    display?: string;
    x?: number;
    y?: number;
} & SchemaNestingBase<SchemaDescriptionTag | SchemaExitTag | SchemaFeatureTag | SchemaNameTag>

export type SchemaFeatureTag = {
    tag: 'Feature';
    key: string;
    global: boolean;
}& SchemaNestingBase<SchemaDescriptionTag | SchemaNameTag>

export type SchemaMapTag = {
    tag: 'Map';
    key: string;
} & SchemaNestingBase<SchemaExitTag | SchemaImageTag | SchemaRoomTag>

export type SchemaStringTag = {
    tag: 'String';
    value: string;
}

export type SchemaTag = SchemaAssetTag |
    SchemaStoryTag |
    SchemaCharacterTag |
    SchemaImageTag |
    SchemaVariableTag |
    SchemaComputedTag |
    SchemaActionTag |
    SchemaUseTag |
    SchemaImportTag |
    SchemaConditionTag |
    SchemaExitTag |
    SchemaDescriptionTag |
    SchemaLineBreakTag |
    SchemaLinkTag |
    SchemaNameTag |
    SchemaRoomTag |
    SchemaFeatureTag |
    SchemaMapTag |
    SchemaStringTag

export class SchemaException extends Error {
    parseTag: ParseTag;
    constructor(message: string, parseTag: ParseTag) {
        super(message)
        this.name = 'WMLSchemaException'
        this.parseTag = parseTag
    }
}

export const isSchemaString = (value: SchemaTag): value is SchemaStringTag => (value.tag === 'String')
