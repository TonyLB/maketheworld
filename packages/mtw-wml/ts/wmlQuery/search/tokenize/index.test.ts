import tokenizer from '.'
import { TokenizeException } from '../../../parser/tokenizer/baseClasses'
import SourceStream from '../../../parser/tokenizer/sourceStream'

describe('wmlQuery select tokenizer', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
    })

    it('should return empty list on empty source', () => {
        const testStream = new SourceStream('')
        expect(tokenizer(testStream)).toEqual([])
    })
    it('should tokenize a single tag', () => {
        const testStream = new SourceStream('Room')
        expect(tokenizer(testStream)).toMatchSnapshot()
    })
    it('should tokenize a sequence of tags', () => {
        const testStream = new SourceStream('Asset Room Exit')
        expect(tokenizer(testStream)).toMatchSnapshot()
    })
    it('should tokenize a complex sequence', () => {
        const testStream = new SourceStream('(Asset, Story) Room:first Exit[to="ABC"]')
        expect(tokenizer(testStream)).toMatchSnapshot()
    })

})
