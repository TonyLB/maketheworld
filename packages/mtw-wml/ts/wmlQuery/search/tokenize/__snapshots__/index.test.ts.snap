// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wmlQuery select tokenizer should tokenize a complex sequence 1`] = `
Array [
  Object {
    "endIdx": 0,
    "startIdx": 0,
    "type": "GroupOpen",
  },
  Object {
    "endIdx": 5,
    "startIdx": 1,
    "tag": "Asset",
    "type": "Tag",
  },
  Object {
    "endIdx": 6,
    "startIdx": 6,
    "type": "Comma",
  },
  Object {
    "endIdx": 12,
    "startIdx": 8,
    "tag": "Story",
    "type": "Tag",
  },
  Object {
    "endIdx": 13,
    "startIdx": 13,
    "type": "GroupClose",
  },
  Object {
    "endIdx": 18,
    "startIdx": 15,
    "tag": "Room",
    "type": "Tag",
  },
  Object {
    "endIdx": 24,
    "startIdx": 19,
    "type": "First",
  },
  Object {
    "endIdx": 29,
    "startIdx": 26,
    "tag": "Exit",
    "type": "Tag",
  },
  Object {
    "endIdx": 39,
    "key": "to",
    "startIdx": 30,
    "type": "Property",
    "value": "ABC",
  },
]
`;

exports[`wmlQuery select tokenizer should tokenize a sequence of tags 1`] = `
Array [
  Object {
    "endIdx": 4,
    "startIdx": 0,
    "tag": "Asset",
    "type": "Tag",
  },
  Object {
    "endIdx": 9,
    "startIdx": 6,
    "tag": "Room",
    "type": "Tag",
  },
  Object {
    "endIdx": 14,
    "startIdx": 11,
    "tag": "Exit",
    "type": "Tag",
  },
]
`;

exports[`wmlQuery select tokenizer should tokenize a single tag 1`] = `
Array [
  Object {
    "endIdx": 3,
    "startIdx": 0,
    "tag": "Room",
    "type": "Tag",
  },
]
`;
